units/db_cleanup_general.go

```diff
@@ -0,0 +1,178 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/model/testresult"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/sometimes"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	dbCleanupJobName = "db-cleanup"
+	cleanupBatchSize = 100 * 1000
+)
+
+func init() {
+	registry.AddJobType(dbCleanupJobName, func() amboy.Job {
+		return makeDbCleanupJob()
+	})
+}
+
+type dataCleanup struct {
```

Comment:
Rather than multiplex all the different data cleanup jobs into 1 job, can we instead keep them as 3 separate jobs and just embed a common job object in all of them, and dependency inject the relevant TTL cleanup function? Muxing them all into 1 implementation makes them hard to change in the future (e.g. if one needs to have some special behavior that's not shared with the other ones).

I would do something like this to inject the relevant deletion logic while still re-using all the other common logic:
```go
type DataCleanupJobBase struct {
    CollectionName string `bson:"collection_name"`
}

func (b *DataCleanupJobBase) runWithDeleteFn(ctx context.Context, deleteFn func(ctx context.Context, env evergreen.Environment, ts time.Time, batchSize int)) {
    // ... all the logic common to all the Run methods
    numDeleted, err := deleteFn(ctx, j.env, ts, batchSize)
    // ... all the remaining logic common to all the Run methods
}

type testLogCleanupJob struct {
    job.Base `bson:"metadata"`
    DataCleanup DataCleanupJobBase `bson:"data_cleanup"`
}

func (j *testLogCleanupJob) Run(ctx context.Context) {
    // do whatever initial job populating logic is necessary...

    j.DataCleanup.runWithDeleteFn(ctx, model.DeleteTestLogsWithLimit)
}
```

Reply:
I agree the if/else case switching is inelegant, and may present problems. I think the best solution is to define the filters where the function is called, and pass them in. This should provide more abstraction without requiring largely identical deletion functions

Reply:
The problem is that you can't serialize a function object into BSON, so when it round-trips through the DB, the TTLFilter will be ignored. I think my proposal doesn't have any copy-pasted code - if you want, you can consolidate the deletion functions by passing the `FilterFunc` to the common base job as I originally suggested.